lib/app.d.ts

```diff
@@ -0,0 +1,244 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+declare namespace Realm {
+    /**
+     * The constructor of MongoDB Realm App.
+     */
+    type AppConstructor = new <FunctionsFactoryType extends FunctionsFactory>(
+        idOrConfiguration: string | AppConfiguration,
+    ) => App<FunctionsFactoryType>;
+
+    /**
+     * A MongoDB Realm App.
+     */
+    interface App<
+        FunctionsFactoryType extends FunctionsFactory = FunctionsFactory
```

Comment:
I find this generic name too verbose.  The problem will become obvious if there are multiple generic parameters. I think the best practice is to use short generic parameter names. What's wrong with `App<F extends FunctionFactory>`

Reply:
The only thing wrong with `F` in my mind is that it's not descriptive.
To me the longer name is more descriptive (hence more verbose) - see https://stackoverflow.com/a/30146204.

Reply:
ok I am convinced. I would be happier if this was `App<TFunctionFactory extends FunctionFactory>` but your choice seems ok.